File: cpp/adaptivekskipmrr.cpp
#include "MyBlas.h"

using namespace std;

template<typename T>
vector<T> mrr(const Matrix &A, const vector<T> &b, const double epsilon, const int k);

int main() {
  Matrix A = {
		{5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0, 0.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0, 2.0},
		{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.0, 5.0}
	};
	vector<double> b = {3.0, 1.0, 4.0, 0.0, 5.0, -1.0, 6.0, -2.0, 7.0, -15.0};
  const double epsilon = 1e-8;
  int k = 0;

  auto x = adaptivekskipmrr(A, b, epsilon, k);
  for(auto e: x) cout << e << endl;
}

template<typename T>
vector<T> adaptivekskipmrr(const Matrix &A, const vector<T> &b, const double epsilon, int k){
  const ul N = A.size();
  const ul max_iter = N*2;
  const double b_norm = vecvec(b, b);
	vector<T> x(N, 0.0);
  vector<T> pre_x(N, 0.0);

  // 初期化
  vv Ar(k + 3, vector<double>(N));
  vv Ay(k + 2, vector<double>(N));
  vector<double> alpha(2*k+3);
  vector<double> beta(2*k+2);
  vector<double> delta(2*k+1);
  beta[0] = 0;

  // 初期残差
  Ar[0] = b - matvec(A, x);
  double residual = vecvec(Ar[0], Ar[0]) / b_norm;
  double pre_residual = residual;

  // 初期反復
  Ar[1] = matvec(A, Ar[0]);
  double zeta = vecvec(Ar[0], Ar[1]) / vecvec(Ar[1], Ar[1]);
  Ay[0] = zeta * Ar[1];
  vector<double> z = -zeta * Ar[0];
  Ar[0] = Ar[0] - Ay[0];
  x = x - z;
  ui i = 1;
  ui index = 1;

  // 反復計算
  while (i < max_iter) {
    // 収束判定
    residual = vecvec(Ar[0], Ar[0]) / b_norm;
    if (residual > pre_residual) {
      // 残差と解を直前の状態に戻す
      x = pre_x;
      Ar[0] = b - matvec(A, x);
      Ar[1] = matvec(A, Ar[0]);
      rAr = vecvec(Ar[0], Ar[1]);
      ArAr = vecvec(Ar[1], Ar[1]);
      zeta = rAr / ArAr;
      Ay[0] = zeta * Ar[1];
      z = -zeta * Ar[0];
      Ar[0] = Ar[0] - Ay[0];
      x = x - z;

      i++;
      index++;
      residual = norm(Ar[0], Ar[0]) / b_norm
      
      // kを1下げる
      if k > 1:
          k -= 1
    } else {
      pre_residual = residual 
      pre_x = x
    }

    if (residual < epsilon) break;

    // 事前計算
    int jj;
    for (int j=1;j<k+2;j++) Ar[j]=matvec(A, Ar[j-1]);
    for (int j=1;j<k+1;j++) Ay[j]=matvec(A, Ay[j-1]);
    for (int j=0;j<2*k+3;j++) { 
      jj = j / 2;
      alpha[j] = vecvec(Ar[jj], Ar[jj + j % 2]);
    }
    for (int j=1;j<2*k+2;j++) { 
      jj = j / 2;
      beta[j] = vecvec(Ay[jj], Ar[jj + j % 2]);
    }
    for (int j=0;j<2*k+1;j++) { 
      jj = j / 2;
      delta[j] = vecvec(Ay[jj], Ay[jj + j % 2]);
    }

    // MrRでの1反復(解の更新)
    double d = alpha[2] * delta[0] - pow(beta[1],2);
    zeta = alpha[1] * delta[0] / d;
    double eta = -alpha[1] * beta[1] / d;
    Ay[0] = eta * Ay[0] + zeta * Ar[1];
    z = eta * z - zeta * Ar[0];
    Ar[0] = Ar[0] - Ay[0];
    Ar[1] = matvec(A, Ar[0]);
    x = x - z;

    // MrRでのk反復
    for (int j=0;j<k;j++) {
      delta[0] = pow(zeta,2) * alpha[2] + eta * zeta * beta[1];
      alpha[0] = alpha[0] - zeta * alpha[1];
      delta[1] = pow(eta,2) * delta[1] + 2 * eta * zeta * beta[2] + pow(zeta,2) * alpha[3];
      beta[1] = eta * beta[1] + zeta * alpha[2] - delta[1];
      alpha[1] = -beta[1];
      for (int l=2;l<2*(k-j)+1;l++) {
        delta[l] = pow(eta,2) * delta[l] + 2 * eta * zeta * beta[l+1] + pow(zeta,2) * alpha[l + 2];
        double tau = eta * beta[l] + zeta * alpha[l + 1];
        beta[l] = tau - delta[l];
        alpha[l] = alpha[l] - tau + beta[l];
      }
      // 解の更新
      d = alpha[2] * delta[0] - pow(beta[1],2);
      zeta = alpha[1] * delta[0] / d;
      eta = -alpha[1] * beta[1] / d;
      Ay[0] = eta * Ay[0] + zeta * Ar[1];
      z = eta * z - zeta * Ar[0];
      Ar[0] = Ar[0] - Ay[0];
      Ar[1] = matvec(A, Ar[0]);
      x = x - z;
    }

    i += (k + 1);
    index++;
  }
  
  return x;
}

File: v2/cpu/mpi/adaptivekskipmrr.py
import numpy as np
from numpy import dot
from numpy.linalg import norm

from .common import start, finish, init, init_mpi


def adaptivekskipmrr(A, b, epsilon, k, T):
    # 共通初期化
    comm, rank, num_of_process = init_mpi()
    local_A, b, x, b_norm, N, max_iter, residual, num_of_solution_updates = init(A, b, T, rank, num_of_process)
    Ax = np.zeros(N, T)

    # 初期化
    Ar = np.zeros((k + 2, N), T)
    Ay = np.zeros((k + 1, N), T)
    rAr = np.zeros(1, T)
    ArAr = np.zeros(1, T)
    alpha = np.zeros(2*k + 3, T)
    beta = np.zeros(2*k + 2, T)
    delta = np.zeros(2*k + 1, T)

    # kの履歴
    k_history = np.zeros(max_iter+1, np.int)
    k_history[0] = k

    # 初期残差
    comm.Allgather(local_A.dot(x), Ax)
    Ar[0] = b - Ax
    residual[0] = norm(Ar[0]) / b_norm

    # 残差減少判定変数
    cur_residual = residual[0].copy()
    pre_residual = residual[0].copy()

    # 初期反復
    if rank == 0:
        start_time = start(method_name='Adaptive k-skip MrR + MPI', k=k)
    comm.Allgather(local_A.dot(Ar[0]), Ar[1])
    rAr = dot(Ar[0], Ar[1])
    ArAr = dot(Ar[1], Ar[1])

    zeta = rAr / ArAr
    Ay[0] = zeta * Ar[1]
    z = -zeta * Ar[0]
    Ar[0] -= Ay[0]
    x -= z

    i = 1
    index = 1
    num_of_solution_updates[1] = 1
    k_history[1] = k

    # 反復計算
    while i < max_iter:
        pre_residual = cur_residual
        cur_residual = norm(Ar[0]) / b_norm
        residual[index] = cur_residual

        # 残差減少判定
        if cur_residual > pre_residual:
            # 解と残差を再計算
            x = pre_x.copy()

            comm.Allgather(local_A.dot(x), Ax)
            comm.Allgather(local_A.dot(Ar[0]), Ar[1])
            rAr = dot(Ar[0], Ar[1])
            ArAr = dot(Ar[1], Ar[1])

            zeta = rAr / ArAr
            Ay[0] = zeta * Ar[1]
            z = -zeta * Ar[0]
            Ar[0] -= Ay[0]
            x -= z

            i += 1
            index += 1
            num_of_solution_updates[index] = i
            residual[index] = norm(Ar[0]) / b_norm

            # kを下げて収束を安定化させる
            if k > 1:
                k -= 1
            k_history[index] = k
        else:
            pre_x = x.copy()

        # 収束判定
        isConverged = cur_residual < epsilon
        if isConverged:
            break

        # 基底計算
        for j in range(1, k + 1):
            comm.Allgather(local_A.dot(Ar[j-1]), Ar[j])
            comm.Allgather(local_A.dot(Ay[j-1]), Ay[j])
        comm.Allgather(local_A.dot(Ar[k]), Ar[k+1])

        # 係数計算
        for j in range(2 * k + 3):
            jj = j // 2
            alpha[j] = dot(Ar[jj], Ar[jj + j % 2])
        for j in range(1, 2 * k + 2):
            jj = j//2
            beta[j] = dot(Ay[jj], Ar[jj + j % 2])
        for j in range(2 * k + 1):
            jj = j // 2
            delta[j] = dot(Ay[jj], Ay[jj + j % 2])

        # MrRでの1反復(解と残差の更新)
        d = alpha[2] * delta[0] - beta[1] ** 2
        zeta = alpha[1] * delta[0] / d
        eta = -alpha[1] * beta[1] / d
        Ay[0] = eta * Ay[0] + zeta * Ar[1]
        z = eta * z - zeta * Ar[0]
        Ar[0] -= Ay[0]
        x -= z

        # MrRでのk反復
        for j in range(k):
            zz = zeta ** 2
            ee = eta ** 2
            ez = eta * zeta
            delta[0] = zz * alpha[2] + ez * beta[1]
            alpha[0] -= zeta * alpha[1]
            delta[1] = ee * delta[1] + 2 * eta * zeta * beta[2] + zz * alpha[3]
            beta[1] = eta * beta[1] + zeta * alpha[2] - delta[1]
            alpha[1] = -beta[1]
            for l in range(2, 2 * (k - j) + 1):
                delta[l] = ee * delta[l] + 2 * ez * \
                    beta[l+1] + zz * alpha[l + 2]
                tau = eta * beta[l] + zeta * alpha[l + 1]
                beta[l] = tau - delta[l]
                alpha[l] -= tau + beta[l]
            # 解と残差の更新
            d = alpha[2] * delta[0] - beta[1] ** 2
            zeta = alpha[1] * delta[0] / d
            eta = -alpha[1] * beta[1] / d
            comm.Allgather(local_A.dot(Ar[0]), Ar[1])
            Ay[0] = eta * Ay[0] + zeta * Ar[1]
            z = eta * z - zeta * Ar[0]
            Ar[0] -= Ay[0]
            x -= z

        i += (k + 1)
        index += 1
        num_of_solution_updates[index] = i
        k_history[index] = k
    else:
        isConverged = False
        residual[index] = norm(Ar[0]) / b_norm

    num_of_iter = i
    if rank == 0:
        elapsed_time = finish(start_time, isConverged,
                              num_of_iter, residual[index], k)
        return elapsed_time, num_of_solution_updates[:index+1], residual[:index+1], k_history[:index+1]
    else:
        exit(0)
